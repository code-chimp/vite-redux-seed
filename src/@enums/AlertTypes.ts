enum AlertTypes {
  Error = 'danger',
  Info = 'info',
  Success = 'success',
  Warning = 'warning',
}

export default AlertTypes;
