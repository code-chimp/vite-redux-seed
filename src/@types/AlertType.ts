import AlertTypes from '../@enums/AlertTypes';

type AlertType = AlertTypes.Error | AlertTypes.Info | AlertTypes.Success | AlertTypes.Warning;

export default AlertType;
