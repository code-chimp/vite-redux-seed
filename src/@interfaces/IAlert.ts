import AlertType from '../@types/AlertType';

interface IAlert {
  id: string;
  type: AlertType;
  text?: string;
  title?: string;
}

export default IAlert;
