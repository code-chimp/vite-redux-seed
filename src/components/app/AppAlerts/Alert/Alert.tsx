import React, { FC, useEffect, useRef } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { IconDefinition } from '@fortawesome/free-regular-svg-icons';
import {
  faCircleCheck,
  faCircleInfo,
  faCircleQuestion,
  faCircleXmark,
  faTriangleExclamation,
} from '@fortawesome/free-solid-svg-icons';
import AlertTypes from '../../../../@enums/AlertTypes';
import IAlert from '../../../../@interfaces/IAlert';
import { removeAlert } from '../../../../store/slices/alerts';
import { useAppDispatch } from '../../../../helpers';

export interface IAlertProps {
  alert: IAlert;
}

const Alert: FC<IAlertProps> = ({ alert }) => {
  const ref = useRef<HTMLDivElement>(null);
  const dispatch = useAppDispatch();

  useEffect(() => {
    const handleClose = () => {
      /* third party functionality not executed by Jest */
      /* istanbul ignore next */
      dispatch(removeAlert(alert.id));
    };
    const el = ref.current;

    el!.addEventListener('close.bs.alert', handleClose);

    return () => {
      el!.removeEventListener('close.bs.alert', handleClose);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  let icon: IconDefinition;

  switch (alert.type) {
    case AlertTypes.Error:
      icon = faCircleXmark;
      break;
    case AlertTypes.Success:
      icon = faCircleCheck;
      break;
    case AlertTypes.Warning:
      icon = faTriangleExclamation;
      break;
    case AlertTypes.Info:
      icon = faCircleInfo;
      break;
    default:
      icon = faCircleQuestion;
  }

  return (
    <div
      ref={ref}
      data-testid={`alert-${alert.id}`}
      className={`alert alert-${alert.type} alert-dismissible fade show d-flex align-items-center`}
      role="alert">
      <FontAwesomeIcon
        data-testid={`alert-${alert.id}-icon`}
        className="flex-shrink-0 me-2"
        icon={icon}
      />
      <div>
        {alert.title ? (
          <h5 data-testid={`alert-${alert.id}-title`} className="mb-0">
            {alert.title}
          </h5>
        ) : null}
        <span data-testid={`alert-${alert.id}-text`}>{alert.text}</span>
      </div>
      <button
        type="button"
        data-testid={`alert-${alert.id}-close`}
        className="btn-close"
        data-bs-dismiss="alert"
        aria-label="Close"></button>
    </div>
  );
};

export default Alert;
