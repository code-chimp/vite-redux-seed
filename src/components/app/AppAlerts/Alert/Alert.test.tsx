/* eslint-disable @typescript-eslint/no-non-null-assertion */
import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import configureMockStore, { MockStore } from 'redux-mock-store';
import { Provider } from 'react-redux';
import Alert from './Alert';
import IAlert from '../../../../@interfaces/IAlert';
import AlertTypes from '../../../../@enums/AlertTypes';
import AlertType from '../../../../@types/AlertType';

const mockStore = configureMockStore([]);
let store: MockStore;

const initialState: Array<IAlert> = [
  {
    id: 'foo',
    text: 'you broke something',
    title: 'naughty dev',
    type: AlertTypes.Error,
  },
  {
    id: 'bar',
    text: 'all the things are fragile',
    type: AlertTypes.Info,
  },
  {
    id: 'baz',
    text: 'you stopped breaking things for the day',
    type: AlertTypes.Success,
  },
  {
    id: 'fibbity',
    text: 'you are about to break it',
    type: AlertTypes.Warning,
  },
  {
    id: 'grag',
    text: 'this really should not work but hey',
    type: 'secondary' as AlertType,
  },
];

describe('components / app / AppAlerts / Alert', () => {
  it('should match the snapshot', () => {
    store = mockStore({ alerts: initialState });
    const { asFragment } = render(
      <Provider store={store}>
        <Alert alert={initialState[0]} />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render a title when provided', () => {
    store = mockStore({ alerts: initialState });
    const target = initialState[0];

    render(
      <Provider store={store}>
        <Alert alert={target} />
      </Provider>,
    );

    const text = screen.getByTestId(`alert-${target.id}-text`);
    const title = screen.getByTestId(`alert-${target.id}-title`);

    expect(text).not.toBeNull();
    expect(text.textContent).toEqual(target.text);
    expect(title).not.toBeNull();
    expect(title.textContent).toEqual(target.title);
  });

  it('should not render a title when not provided', () => {
    store = mockStore({ alerts: initialState });
    const target = initialState[1];

    render(
      <Provider store={store}>
        <Alert alert={target} />
      </Provider>,
    );

    const text = screen.getByTestId(`alert-${target.id}-text`);
    const title = screen.queryByTestId(`alert-${target.id}-title`);

    expect(text).not.toBeNull();
    expect(text.textContent).toEqual(target.text);
    expect(title).not.toBeInTheDocument();
  });

  it('should render an error alert with the appropriate icon', () => {
    store = mockStore({ alerts: initialState });
    const target = initialState[0];

    render(
      <Provider store={store}>
        <Alert alert={target} />
      </Provider>,
    );

    const icon = screen.getByTestId(`alert-${target.id}-icon`);

    expect(icon).not.toBeNull();
    expect(icon!.classList.contains('fa-circle-xmark')).toBe(true);
  });

  it('should render an info alert with the appropriate icon', () => {
    store = mockStore({ alerts: initialState });
    const target = initialState[1];

    render(
      <Provider store={store}>
        <Alert alert={target} />
      </Provider>,
    );

    const icon = screen.getByTestId(`alert-${target.id}-icon`);

    expect(icon).not.toBeNull();
    expect(icon!.classList.contains('fa-circle-info')).toBe(true);
  });

  it('should render a success alert with the appropriate icon', () => {
    store = mockStore({ alerts: initialState });
    const target = initialState[2];

    render(
      <Provider store={store}>
        <Alert alert={target} />
      </Provider>,
    );

    const icon = screen.getByTestId(`alert-${target.id}-icon`);

    expect(icon).not.toBeNull();
    expect(icon!.classList.contains('fa-circle-check')).toBe(true);
  });

  it('should render a warning alert with the appropriate icon', () => {
    store = mockStore({ alerts: initialState });
    // eslint-disable-next-line no-magic-numbers
    const target = initialState[3];

    render(
      <Provider store={store}>
        <Alert alert={target} />
      </Provider>,
    );

    const icon = screen.getByTestId(`alert-${target.id}-icon`);

    expect(icon).not.toBeNull();
    expect(icon!.classList.contains('fa-triangle-exclamation')).toBe(true);
  });

  it('should render a question mark icon if passed an unknown AlertType', () => {
    store = mockStore({ alerts: initialState });
    // eslint-disable-next-line no-magic-numbers
    const target = initialState[4];

    render(
      <Provider store={store}>
        <Alert alert={target} />
      </Provider>,
    );

    const icon = screen.getByTestId(`alert-${target.id}-icon`);

    expect(icon).not.toBeNull();
    expect(icon!.classList.contains('fa-circle-question')).toBe(true);
  });

  it('should dispatch an action when close button is clicked', () => {
    store = mockStore({ alerts: initialState });
    const target = initialState[1];

    render(
      <Provider store={store}>
        <Alert alert={target} />
      </Provider>,
    );
    // const expected = [
    //   {
    //     type: alerts.actions.removeAlert.type,
    //     payload: initialState[0].id,
    //   },
    // ];

    const closeButton = screen.getByTestId(`alert-${target.id}-close`);
    expect(closeButton).not.toBeNull();

    fireEvent.click(closeButton!);

    // NOTE: the actual action is not getting dispatched as it depends on the
    //       Bootstrap JavaScript which is not being executed by the test framework
    // expect(store.getActions()).toEqual(expected);
    expect(store.getActions()).toEqual([]);
  });
});
