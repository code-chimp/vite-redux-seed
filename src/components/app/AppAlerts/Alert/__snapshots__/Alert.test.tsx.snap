// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components / app / AppAlerts / Alert > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="alert alert-danger alert-dismissible fade show d-flex align-items-center"
    data-testid="alert-foo"
    role="alert"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-circle-xmark flex-shrink-0 me-2"
      data-icon="circle-xmark"
      data-prefix="fas"
      data-testid="alert-foo-icon"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM175 175c9.4-9.4 24.6-9.4 33.9 0l47 47 47-47c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9l-47 47 47 47c9.4 9.4 9.4 24.6 0 33.9s-24.6 9.4-33.9 0l-47-47-47 47c-9.4 9.4-24.6 9.4-33.9 0s-9.4-24.6 0-33.9l47-47-47-47c-9.4-9.4-9.4-24.6 0-33.9z"
        fill="currentColor"
      />
    </svg>
    <div>
      <h5
        class="mb-0"
        data-testid="alert-foo-title"
      >
        naughty dev
      </h5>
      <span
        data-testid="alert-foo-text"
      >
        you broke something
      </span>
    </div>
    <button
      aria-label="Close"
      class="btn-close"
      data-bs-dismiss="alert"
      data-testid="alert-foo-close"
      type="button"
    />
  </div>
</DocumentFragment>
`;
